[DO_NOT_MERGE] Allow specyfing ImageStream, ImageStreamTag and DockerImage in BuildOutput's To field

soltysh opened the pull request:
Primary goal is to fix #2443. Additionally I've addressed issues mentioned in #2475 regarding build validation (iow. #2586). 

Currently this is waiting for #2475 and #2486 to be merged first.

@bparees @mfojtik PTAL

mfojtik commented on examples/sample-app/application-template-custombuild.json:
```diff
@@ -66,6 +66,7 @@
       "parameters": {
         "output": {
           "to": {
+            "kind": "ImageStream",
```
why you have to specify this? i thought the default here is ImageStream

soltysh replied:
It's not, we should then, analogically to what we have in From, where ImageStreamTag is the default (see [here](https://github.com/openshift/origin/blob/master/pkg/build/api/v1/conversion.go#L16-L30)).

mfojtik replied:
yeah, in both cases I think we agreed that you don't have to specify the Kind and we assume that the Name is the 'default' Kind. From and To has different defaults, right? @bparees

soltysh replied:
Right, we need to make it clear, and have it in code explicitly defined.

liggitt replied:
does the default get populated so that API consumers don't have to know that "" means "ImageStream"?

bparees commented on pkg/build/api/validation/validation.go:
```diff
@@ -128,24 +136,20 @@ func validateRevision(revision *buildapi.SourceRevision) fielderrors.ValidationE
-		if len(kind) == 0 {
-			kind = "ImageStream"
-			output.To.Kind = kind
+		if kind != "ImageStream" && kind != "ImageStreamTag" && kind != "DockerImage" {
```
Kind should be allowed to be empty, and if empty, it means ImageStream.  Validation might not be the right place to enforce that, but i don't think you should remove the logic entirely as you appear to have done.

soltysh replied:
Added in defaults.

bparees commented on pkg/build/api/validation/validation.go:
```diff
@@ -213,13 +205,33 @@ func validateSourceStrategy(strategy *buildapi.SourceBuildStrategy) fielderrors.
 
+func validateStrategyFrom(from *kapi.ObjectReference) fielderrors.ValidationErrorList {
+	allErrs := fielderrors.ValidationErrorList{}
+	if from.Kind != "ImageStreamImage" && from.Kind != "ImageStreamTag" && from.Kind != "DockerImage" {
```
again, empty Kind means imagestream, we need to tolerate that either here, or through defaulting.

soltysh replied:
Actually from what I see in the defaulting funcs it means ImageStreamTag for from (see [here](https://github.com/openshift/origin/blob/master/pkg/build/api/v1beta3/conversion.go#L16-L30)). Should it be IS?

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -85,20 +87,16 @@ func (bc *BuildController) nextBuildStatus(build *buildapi.Build) error {
 
-		repo, err := bc.ImageStreamClient.GetImageStream(namespace, ref.Name)
+		var err error
+		spec, err = bc.resolveImageStreamReference(ref, namespace)
```
is there a good reason we didn't resolve all this when we created the build object, instead of waiting until now to do it?

bparees commented on pkg/build/controller/controller.go:
```diff
@@ -134,6 +132,40 @@ func (bc *BuildController) nextBuildStatus(build *buildapi.Build) error {
 
+// resolveImageStreamReference looks up the ImageStream[Tag/Image] and converts it to a
+// docker pull spec that can be used in an Image field.
+func (bc *BuildController) resolveImageStreamReference(to *kapi.ObjectReference, namespace string) (string, error) {
```
is this different from the version in generator.go?  can we unify them?

soltysh replied:
Slightly different, as at the time of writing From used IST, ISI, DI, whereas To used IS, IST, DI. Although I'm not sure why we don't allow IS in the From field, but we can't allow having ISI in the To field.

bparees commented:
empty kind means ImageStream, for historical reasons.

liggitt commented:
That should definitely go in defaults so every client doesn't have to know that also when interpreting data that comes from the API

liggitt commented on pkg/build/api/validation/validation.go:
```diff
@@ -44,11 +44,19 @@ func ValidateBuildConfig(config *buildapi.BuildConfig) fielderrors.ValidationErr
+	for i, trg := range config.Triggers {
+		allErrs = append(allErrs, validateTrigger(&trg).PrefixIndex(i).Prefix("triggers")...)
+		if trg.Type == buildapi.ImageChangeBuildTriggerType {
+			ictCount++
```
A deployment can't trigger on different image changes? What if I have a deployment that has two containers, and I want a new deployment when either of them changes?

liggitt replied:
Oops... this is builds, not deployments. Too many triggers, too little brain...

soltysh commented:
I've addressed all the comments:
- currently for both To and From the default is ImageStreamTag applied at defaulting functions
- moved To verification to generator similarly to where From is
- squashed this into one, as I got some weird problems with validation that I didn't want to deal in separate commits

I'm currently fighting with conversions tests, but I think I'll leave it for tomorrow.

bparees commented on pkg/build/generator/generator.go:
```diff
@@ -267,60 +267,64 @@ func (g *BuildGenerator) generateBuildFromConfig(ctx kapi.Context, bc *buildapi.
-			return "", err
-		}
-		return image.String(), nil
+	glog.V(4).Infof("Resolving ImageStreamReference %s of Kind %s in namespace %s", ref.Name, ref.Kind, namespace)
```
looking at the conversions  and talking to @smarterclayton we are still using ImageStream as the internal Kind for a To field.  So i don't think you can remove that case from this switch.

I think we need to discuss this PR more also because @smarterclayton says he had already fixed To to support ImageStreamTag and DockerImage.

soltysh commented:
Sadly informing the patient did not make through the surgery....

soltysh commented:
Trelloed here for future https://trello.com/c/xrPu7upe/590-clean-validation-in-builds-future-build
